java/com/google/flatbuffers/FlexBuffersBuilder.java

```diff
@@ -0,0 +1,981 @@
+            bb.putFloat((float) val);
+        } else if (byteWidth == 8) {
+            bb.putDouble(val);
+        } else {
+            throw new FlexBuffers.FlexBufferException(String.format("Invalid byte width of %d", byteWidth));
+        }
+    }
+
+    private int putMap(String key, MapBuilderCallback vbc) {
+        int start = stack.size();
+        vbc.run(new MapBuilder(this));
+        return endMap(key, start);
+    }
+
+    private int endMap(String key, int start) {
+        Value vKey = putKey(key);
+
+        // Make sure keys are all strings:
+        for (int i = start; i < stack.size(); i++) {
+            assert (stack.get(i) != null && stack.get(i).key.type == FBT_KEY);
+        }
+
+        Collections.sort(stack.subList(start, stack.size()));
+
+        Value keys = createKeyVector(start, stack.size() - start);
+        Value vec = createVector(start, stack.size() - start, false, false, keys);
+        // Remove temp elements and return map.
+        while (stack.size() > start) {
+            stack.remove(stack.size() - 1);
+        }
+        stack.add(new KeyValue(vKey, vec));
+        return (int) vec.iValue;
+    }
+
+    private Value createKeyVector(int start, int length) {
+        // Figure out smallest bit width we can store this vector with.
+        BitWidth bitWidth = BitWidth.max(WIDTH_8, widthUInBits(length));
+        int prefixElems = 1;
+        int vectorType = FBT_KEY;
+        // Check bit widths and types for all elements.
+        for (int i = start; i < stack.size(); i++) {
+            BitWidth elemWidth = stack.get(i).key.elemWidth(bb.position(), i + prefixElems);
+            bitWidth = BitWidth.max(bitWidth, elemWidth);
+            if (i == start) {
+                vectorType = stack.get(i).key.type;
+            } else {
+                // If you get this assert, you are writing a typed vector with
+                // elements that are not all the same type.
+                assert (vectorType == stack.get(i).key.type);
+            }
+        }
+
+        int byteWidth = align(bitWidth);
+        // Write vector. First the keys width/offset if available, and size.
+        writeInt(length, byteWidth);
+        // Then the actual data.
+        int vloc = bb.position();
+        for (int i = start; i < stack.size(); i++) {
+            writeAny(stack.get(i).key, byteWidth);
+        }
+        // Then the types.
+        return new Value(FlexBuffers.toTypedVector(vectorType,0), bitWidth, vloc);
+    }
+
+    /**
+     * Callback interface used to add elements to a vector
+     */
+    public interface VectorBuilderCallback {
+        void run(VectorBuilder vb);
+    }
+
+    /**
+     * Callback interface used to add elements to a map
+     */
+    public interface MapBuilderCallback {
+        void run(MapBuilder vb);
+    }
+
+    /**
+     * This class will be use to construct vectors in a FlexBuffer message
+     */
+    public static class VectorBuilder {
```

Comment:
Why do we need this helper class?

Reply:
Just to strict api while you a inserting a vector. So no call to `putInt("myKey", 10);` inside a vector. Or `putInt(10)` inside a map.

Reply:
C++ doesn't necessarily protect against this either. Again, I'd prefer to favor efficiency where possible.

Reply:
Yes, I removed that interface to keep just the efficient API